flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -412,30 +439,32 @@ private void releaseFailedContainerAndRequestNewContainerIfRequired(ContainerId
 
 		final ResourceID resourceId = new ResourceID(containerId.toString());
 		// release the failed container
-		workerNodeMap.remove(resourceId);
+		YarnWorkerNode yarnWorkerNode = workerNodeMap.remove(resourceId);
 		resourceManagerClient.releaseAssignedContainer(containerId);
 		// and ask for a new one
-		requestYarnContainerIfRequired();
+		requestYarnContainerIfRequired(yarnWorkerNode.getContainer().getResource());
 	}
 
 	private void returnExcessContainer(Container excessContainer) {
 		log.info("Returning excess container {}.", excessContainer.getId());
 		resourceManagerClient.releaseAssignedContainer(excessContainer.getId());
 	}
 
-	private void removeContainerRequest(AMRMClient.ContainerRequest pendingContainerRequest) {
-		numPendingContainerRequests--;
-
-		log.info("Removing container request {}. Pending container requests {}.", pendingContainerRequest, numPendingContainerRequests);
-
+	private void removeContainerRequest(AMRMClient.ContainerRequest pendingContainerRequest, WorkerResourceSpec workerResourceSpec) {
+		log.info("Removing container request {}.", pendingContainerRequest);
+		pendingWorkerCounter.decreaseAndGet(workerResourceSpec);
```

Comment:
What about removing it from `workerSpecContainerResourceAdapter`?

Reply:
Differently asked, when do we clean `workerSpecContainerResourceAdapter` up?

Reply:
Unfortunately, we don't clean `workerSpecContainerResourceAdapter` up.

I think the upper bond of the amount of records really depends on how many different `WorkerResourceSpec` do we have. If we want to clean the unused records up, the `WorkerSpecContainerResourceAdapter` will need `YarnResourceManager` to tell it which WorkerResourceSpec is no longer needed (all corresponding TMs are completed and no pending ones). I'm not sure whether this complexity is necessary.

Reply:
Hmm, would it be possible to have a periodic cleanup task?

Reply:
Not saying that we have to implement it right away. I just want to know how one could fix this as a follow-up task.

Reply:
True, a periodic cleanup task should work there, or we can trigger a clean-up if the number of entries stored in `WorkerSpecContainerResourceAdapter` reaches a certain threshold.

I think the clean-up is definitely doable. It's just not very necessary at the time being.